Improve documentation for pipeline parallelism.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49577 Add benchmark for torch.distributed.pipeline.sync.Pipe
* **#48638 Improve documentation for pipeline parallelism.**
* #48630 Cleanup APIs for pipeline parallelism.
* #48470 Test pipeline parallelism works with DDP.

Polishing up some of the docs for the main `Pipe` class and its
`forward` method.

Differential Revision: [D25237705](https://our.internmc.facebook.com/intern/diff/D25237705/)

pritamdamania committed:
Improve documentation for pipeline parallelism.

Polishing up some of the docs for the main `Pipe` class and its
`forward` method.

Differential Revision: [D25237705](https://our.internmc.facebook.com/intern/diff/D25237705/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/pipeline/sync/pipe.py:
```diff
@@ -137,64 +137,65 @@ def _split_module(modules: nn.Sequential) -> Tuple[List[nn.Sequential], List[tor
-    to train on Pipe_. If the module requires lots of memory, Pipe will be
-    very efficient.
-    ::
+    to train on using synchronous pipeline parallelism. If the module requires lots of memory and doesn't fit on a single GPU, pipeline parallelism is a useful technique to employ for training.
```
nit: shall we break long lines in this file into shorter ones?

mrshenli commented on torch/distributed/pipeline/sync/pipe.py:
```diff
@@ -137,64 +137,65 @@ def _split_module(modules: nn.Sequential) -> Tuple[List[nn.Sequential], List[tor
-            ``'except_last'``, or ``'never'`` (default: ``'except_last'``)
+            ``'except_last'``, or ``'never'`` (default: ``'except_last'``). ``'never'`` disables checkpointing completely, ``'except_last'`` enables checkpointing for all micro-batches except the last one and ``'always'`` enables checkpointing for all micro-batches.
         deferred_batch_norm (bool):
             whether to use deferred BatchNorm moving statistics (default:
```
double quote `BatchNorm`?

mrshenli commented on torch/distributed/pipeline/sync/pipe.py:
```diff
@@ -137,64 +137,65 @@ def _split_module(modules: nn.Sequential) -> Tuple[List[nn.Sequential], List[tor
-
-    Keyword Args:
         chunks (int):
             number of micro-batches (default: ``1``)
```
This is assuming the first dimension is the batch dimension? Do we want to mention what if `chunks` > the samples in a batch?

pritamdamania87 replied:
This clarification probably belongs in the forward method, I've updated those docs to reflect this.

mrshenli commented on torch/distributed/pipeline/sync/pipe.py:
```diff
@@ -137,64 +137,65 @@ def _split_module(modules: nn.Sequential) -> Tuple[List[nn.Sequential], List[tor
+    .. note::
+        You can wrap a :class:`Pipe` model with
+        :class:`torch.nn.parallel.DistributedDataParallel` only when the
+        checkpoint parameter of :class:`Pipe` is ``'never'``.
```
are we going to check this?

pritamdamania87 replied:
No we are not checking this, this is just a note to the user that other modes will not work with DDP.

mrshenli commented on torch/distributed/pipeline/sync/pipe.py:
```diff
@@ -306,21 +307,24 @@ def _ensure_copy_streams(self) -> List[List[AbstractStream]]:
         modify the input and output signature of the underlying module. But
         there's type restriction. Input and output have to be a
         :class:`~torch.Tensor` or a sequence of tensors. This restriction is
         applied at partition boundaries too.
```
Any reason for this limitation? Are we going to drop this in the future?

pritamdamania87 replied:
Its tricky to split an arbitrary python object into multiple microbatches, that is the primary motivation of this limitation.

mrshenli reviewed (approved):
LGTM! Left some minor comments.

pritamdamania committed:
Update on "Improve documentation for pipeline parallelism."


Polishing up some of the docs for the main `Pipe` class and its
`forward` method.

Differential Revision: [D25237705](https://our.internmc.facebook.com/intern/diff/D25237705/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Improve documentation for pipeline parallelism."


Polishing up some of the docs for the main `Pipe` class and its
`forward` method.

Differential Revision: [D25237705](https://our.internmc.facebook.com/intern/diff/D25237705/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Improve documentation for pipeline parallelism."


Polishing up some of the docs for the main `Pipe` class and its
`forward` method.

Differential Revision: [D25237705](https://our.internmc.facebook.com/intern/diff/D25237705/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Improve documentation for pipeline parallelism."


Polishing up some of the docs for the main `Pipe` class and its
`forward` method.

Differential Revision: [D25237705](https://our.internmc.facebook.com/intern/diff/D25237705/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Improve documentation for pipeline parallelism."


Polishing up some of the docs for the main `Pipe` class and its
`forward` method.

Differential Revision: [D25237705](https://our.internmc.facebook.com/intern/diff/D25237705/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Improve documentation for pipeline parallelism."


Polishing up some of the docs for the main `Pipe` class and its
`forward` method.

Differential Revision: [D25237705](https://our.internmc.facebook.com/intern/diff/D25237705/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Improve documentation for pipeline parallelism."


Polishing up some of the docs for the main `Pipe` class and its
`forward` method.

Differential Revision: [D25237705](https://our.internmc.facebook.com/intern/diff/D25237705/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Improve documentation for pipeline parallelism."


Polishing up some of the docs for the main `Pipe` class and its
`forward` method.

Differential Revision: [D25237705](https://our.internmc.facebook.com/intern/diff/D25237705/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Improve documentation for pipeline parallelism."


Polishing up some of the docs for the main `Pipe` class and its
`forward` method.

Differential Revision: [D25237705](https://our.internmc.facebook.com/intern/diff/D25237705/)

[ghstack-poisoned]
